Source/Core/DolphinWX/LuaScriptFrame.cpp

```diff
@@ -0,0 +1,553 @@
+
+  if (dialog->ShowModal() == wxID_CANCEL)
+    return;
+
+  file_path->SetValue(dialog->GetPath());
+  dialog->Destroy();
+}
+
+void LuaScriptFrame::RunOnButtonClick(wxCommandEvent& event)
+{
+  if (lua_thread == nullptr)
+  {
+    lua_thread = new LuaThread(this, file_path->GetValue());
+    lua_thread->Run();
+  }
+}
+
+void LuaScriptFrame::StopOnButtonClick(wxCommandEvent& event)
+{
+  // Kill current Lua thread
+  if (lua_thread != nullptr && lua_thread->IsRunning())
+  {
+    lua_thread->Kill();
+    lua_thread = nullptr;
+  }
+}
+
+// The callback function that tells the emulator what to actually press
+void LuaScriptFrame::GetValues(GCPadStatus* status)
+{
+  // We lock here to prevent an issue where pad_status could be deleted during the
+  // middle of this function's execution
+  std::unique_lock<std::mutex> lock(lua_mutex);
+
+  if (lua_thread == nullptr)
+    return;
+
+  if (pad_status->stickX != GCPadStatus::MAIN_STICK_CENTER_X)
+    status->stickX = pad_status->stickX;
+
+  if (pad_status->stickY != GCPadStatus::MAIN_STICK_CENTER_Y)
+    status->stickY = pad_status->stickY;
+
+  if (pad_status->triggerLeft != 0)
+    status->triggerLeft = pad_status->triggerLeft;
+
+  if (pad_status->triggerRight != 0)
+    status->triggerRight = pad_status->triggerRight;
+
+  if (pad_status->substickX != GCPadStatus::C_STICK_CENTER_X)
+    status->substickX = pad_status->substickX;
+
+  if (pad_status->substickY != GCPadStatus::C_STICK_CENTER_Y)
+    status->substickY = pad_status->substickY;
+
+  status->button |= pad_status->button;
+}
+
+void LuaScriptFrame::NullifyLuaThread()
+{
+  lua_thread = nullptr;
+}
+
+// Sets status to the default values
+void ClearPad(GCPadStatus* status)
+{
+  status->button = 0;
+  status->stickX = GCPadStatus::MAIN_STICK_CENTER_X;
+  status->stickY = GCPadStatus::MAIN_STICK_CENTER_Y;
+  status->triggerLeft = 0;
+  status->triggerRight = 0;
+  status->substickX = GCPadStatus::C_STICK_CENTER_X;
+  status->substickY = GCPadStatus::C_STICK_CENTER_Y;
+}
+
+// Functions to register with Lua
+#pragma region Lua_Functs
+
+// Prints a string to the text control of this frame
+int printToTextCtrl(lua_State* L)
+{
+  currentWindow->Log(lua_tostring(L, 1));
+  currentWindow->Log("\n");
+
+  return 0;
+}
+
+// Steps a frame if the emulator is paused, pauses it otherwise.
+int frameAdvance(lua_State* L)
+{
+  u64 currentFrame = Movie::GetCurrentFrame();
+  Core::DoFrameStep();
+
+  // Block until a frame has actually processed
+  // Prevents a script from executing it's main loop more than once per frame.
+  while (currentFrame == Movie::GetCurrentFrame())
+    ;
+  return 0;
+}
+
+int getFrameCount(lua_State* L)
+{
+  lua_pushinteger(L, Movie::GetCurrentFrame());
+  return 1;
+}
+
+int getMovieLength(lua_State* L)
+{
+  if (Movie::IsMovieActive())
+  {
+    lua_pushinteger(L, Movie::GetTotalFrames());
+  }
+  else
+  {
+    luaL_error(L, "No active movie.");

```

Comment:
Shouldn't this return `0` here since nothing was added to the Lua stack in this case?

Reply:
Yes, returning 0 items / everything nil is vastly preferable to throwing errors.